localstack/services/awslambda/lambda_api.py

```diff
@@ -663,49 +668,115 @@ def process_sns_notification(
             arn = source["FunctionArn"]
             for chunk in chunks(records, source["BatchSize"]):
                 shard_id = "shardId-000000000000"
                 event = {
-                    "Records": [
-                        {
-                            "eventID": "{0}:{1}".format(shard_id, rec["sequenceNumber"]),
-                            "eventSourceARN": stream_arn,
-                            "eventSource": "aws:kinesis",
-                            "eventVersion": "1.0",
-                            "eventName": "aws:kinesis:record",
-                            "invokeIdentityArn": "arn:aws:iam::{0}:role/lambda-role".format(
-                                TEST_AWS_ACCOUNT_ID
-                            ),
-                            "awsRegion": aws_stack.get_region(),
-                            "kinesis": rec,
-                        }
-                        for rec in chunk
-                    ]
+                    "Records": create_records(
+                        shard_id, stream_arn, chunk, is_dynamodb_record=is_dynamodb_record
+                    )
                 }
                 lock_discriminator = None
-                if not config.SYNCHRONOUS_KINESIS_EVENTS:
+                if not is_dynamodb_record and not config.SYNCHRONOUS_KINESIS_EVENTS:
                     lock_discriminator = f"{stream_arn}/{shard_id}"
                     lambda_executors.LAMBDA_ASYNC_LOCKS.assure_lock_present(
                         lock_discriminator, BoundedSemaphore(source["ParallelizationFactor"])
                     )
+
+                def on_failure_callback(
+                    result, func_arn, event, error=None, dlq_sent=None, **kwargs
+                ):
+                    # streams only supports "OnFailure"
+                    if not error:
+                        return
+
+                    details = {
+                        "shardId": shard_id,
+                        "startSequenceNumber": chunk[0]["sequenceNumber"],
+                        "endSequenceNumber": chunk[-1]["sequenceNumber"],
+                        "approximateArrivalOfFirstRecord": timestamp_millis(
+                            chunk[0]["approximateArrivalTimestamp"]
+                        ),
+                        "approximateArrivalOfLastRecord": timestamp_millis(
+                            chunk[-1]["approximateArrivalTimestamp"]
+                        ),
+                        "batchSize": source["BatchSize"],
+                        "streamArn": stream_arn,
+                    }
+                    payload = {
+                        "version": "1.0",
+                        "timestamp": timestamp_millis(),
+                        "requestContext": {
+                            "requestId": long_uid(),
+                            "functionArn": func_arn,
+                            "condition": "RetryAttemptsExhausted",
+                            "approximateInvokeCount": 1,
```

Comment:
You mean `condition` or the `approximateInvokeCount`?
For the `condition`: it's what AWS gives back in that scenario, don't know what else it could be?
For `approximateInvokeCount`: as I understand, we currently invoke the lambda only once, there seems no logic in place in yet, to run lambdas x times?

What would be your expectations for the values?